xonsh/dirstack.py

```diff
@@ -0,0 +1,216 @@
+import os
+import builtins
+from argparse import ArgumentParser
+
+DIRSTACK = []
+
+
+def pushd(args, stdin=None):
+    global DIRSTACK
+
+    try:
+        args = pushd_parser.parse_args(args)
+    except SystemExit:
+        return None, None
+
+    pwd = builtins.__xonsh_env__['PWD']
+
+    if args.dir is None:
+        try:
+            new_pwd = DIRSTACK.pop(0)
+        except:
+            e = 'pushd: Directory stack is empty\n'
+            return None, e
+    elif os.path.isdir(args.dir):
+        new_pwd = args.dir
+    else:
+        try:
+            num = int(args.dir[1:])
+            assert num >= 0
+        except:
+            e = 'Invalid argument to pushd: {0}\n'
+            return None, e.format(args.dir)
+        if num > len(DIRSTACK):
+            e = 'Too few elements in dirstack ({0} elements)\n'
+            return None, e.format(len(DIRSTACK))
+        elif args.dir.startswith('+'):
+            if num == len(DIRSTACK):
+                new_pwd = None
+            else:
+                new_pwd = DIRSTACK.pop(len(DIRSTACK)-1-num)
+        elif args.dir.startswith('-'):
+            if num == 0:
+                new_pwd = None
+            else:
+                new_pwd = DIRSTACK.pop(num-1)
+        else:
+            e = 'Invalid argument to pushd: {0}\n'
+            return None, e.format(args.dir)
+    if new_pwd is not None:
+        DIRSTACK.insert(0, os.path.expanduser(pwd))
+
+        o = None
+        e = None
+        if args.cd:
+            o, e = builtins.default_aliases['cd']([new_pwd], None)
+
+        if e is not None:
+            return None, e
+
+    if not builtins.__xonsh_env__.get('QUIET_PUSHD', False):
+        return dirs([], None)
+
+    return None, None
+
+
+def popd(args, stdin=None):
+    global DIRSTACK
+
+    try:
+        args = pushd_parser.parse_args(args)
+    except SystemExit:
+        return None, None
+
+    if args.dir is None:
+        try:
+            new_pwd = DIRSTACK.pop(0)
+        except:
+            e = 'popd: Directory stack is empty\n'
+            return None, e
+    else:
+        try:
+            num = int(args.dir[1:])
+            assert num >= 0
+        except:
```

Comment:
It is much better to do `if num >= 0` than an except.  Also, I think that int will only ever throw a value error.  So that error is what should be caught.